bin/check-changelog.js

```diff
@@ -0,0 +1,108 @@
+/**
+ * Optionally given a set of file paths as arguments, checks whether all files
+ * modified on the current branch (including those passed as arguments) include
+ * a requisite CHANGELOG entry.
+ */
+
+/**
+ * External dependencies
+ */
+
+const { relative, sep, join } = require( 'path' );
+const minimatch = require( 'minimatch' );
+const { execSync } = require( 'child_process' );
+const { flow, toString, trim } = require( 'lodash' );
+
+/**
+ * Internal dependencies
+ */
+
+const { ignoreChanges = [] } = require( '../lerna' );
+
+/**
+ * Utility function to trim the result of an execSync string result.
+ *
+ * @param {string} command Command to exec.
+ *
+ * @return {string} Result of exec.
+ */
+const exec = flow( execSync, toString, trim );
+
+/**
+ * Returns true if the given path is ignored by the Lerna `ignoreChanges`
+ * configuration, or false otherwise.
+ *
+ * @param {string} path Path to test.
+ *
+ * @return {boolean} Whether path is ignored.
+ */
+function isIgnored( path ) {
+	return ignoreChanges.some( ( pattern ) => minimatch( path, pattern ) );
+}
+
+/**
+ * Merge base commit SHA on which to perform git file diff.
+ *
+ * @type {string}
+ */
+const base = exec( 'git merge-base master HEAD' );
```

Comment:
This part is tricky and might cause a lot of confusion because we reference the local copy of `master` branch. I run into this issue locally where my `master` branch is out of date if you compare with the upstream. Therefore when I run this script I see missing `CHANGELOG.md` files in folders that weren't updated in the branch where I tried to add a new commit. In fact, I wanted to add a new commit directly to this branch to verify how it would work with some flows. In my case `git merge-base origin/master HEAD` solves the issue, but that might differ for forks and can be also changed to something else.

I'm also wondering if this isn't a temporary issue caused by the fact that we don't update changelog files in a way this should happen. My assumption is that it wouldn't be an issue when this change lands and folks update their `master` to the latest version.

As a way to avoid so much confusion maybe we should include the note prompting to update `master` to the latest version and try again?

Reply:
On the up to date `master` branch it works like a charm ✨

Reply:
>This part is tricky and might cause a lot of confusion because we reference the local copy of `master` branch. I run into this issue locally where my `master` branch is out of date if you compare with the upstream.

Argh, I'd hoped to have found some reliable way here. The other trouble too with dealing with branches is that Travis does some funny things with branches, which makes using the `git` command not as reliable ([reference](https://graysonkoonce.com/getting-the-current-branch-name-during-a-pull-request-in-travis-ci/)).

There's seemingly many ways to check for differences; I want to dig to see if there could be another option here, before resorting to the forced update prompt.

Reply:
Jest supports something similar with the following flags:
- [--changedSince](https://jestjs.io/docs/en/cli#changedsince) - "Runs tests related to the changes since the provided branch. If the current branch has diverged from the given branch, then only changes made locally will be tested."
- [--onlyChanged](https://jestjs.io/docs/en/cli#onlychanged) - "Attempts to identify which tests to run based on which files have changed in the current repository."

It might be a nice way to compare the approach they took.